pkg/endpoint/endpoint.go

```diff
@@ -2353,14 +2353,47 @@ func (e *Endpoint) setPolicyRevision(rev uint64) {
 			close(ps.ch)
 			delete(e.policyRevisionSignals, ps)
 		default:
-			if rev >= ps.wantedRev {
+			// We only close the channel when e.policyRevision >= ps.wantedRev
+			// and, if the endpoint has (or had) a proxy set up, when the
+			// e.proxyPolicyRevision >= ps.wantedRev.
+			if e.policyRevision >= ps.wantedRev &&
+				(!e.HasProxy() || e.proxyPolicyRevision >= ps.wantedRev) {
+
 				close(ps.ch)
 				delete(e.policyRevisionSignals, ps)
 			}
 		}
 	}
 }
 
+// setProxyPolicyRevision sets the policy wantedRev with the given revision.
+func (e *Endpoint) setProxyPolicyRevision(rev uint64) {
+	e.proxyPolicyRevision = rev
+	for ps := range e.policyRevisionSignals {
+		select {
+		case <-ps.ctx.Done():
+			close(ps.ch)
+			delete(e.policyRevisionSignals, ps)
+		default:
+			// We only close the channel when e.policyRevision >= ps.wantedRev
+			// and, if the endpoint has (or had) a proxy set up, when the
+			// e.proxyPolicyRevision >= ps.wantedRev.
+			if e.policyRevision >= ps.wantedRev &&
+				(!e.HasProxy() || e.proxyPolicyRevision >= ps.wantedRev) {
+
+				close(ps.ch)
+				delete(e.policyRevisionSignals, ps)
+			}
+		}
+	}
+}
+
+// HasProxy returns true if the given endpoint has a proxy redirect configured.
+// Must be called with e.RLock held
+func (e *Endpoint) HasProxy() bool {
+	return e.DesiredL4Policy.HasRedirect() || e.RealizedL4Policy.HasRedirect()
```

Comment:
That's not reliable. In the sidecar HTTP proxy mode, we have no redirects. See `Endpoint.addNewRedirectsFromMap` in `pkg/endpoint/bpf.go`.

In the sidecar HTTP proxy mode, you need to check whether there are L7 rules in the policy. See how it's done in `XDSServer.UpdateNetworkPolicy` in `pkg/envoy/server.go`: https://github.com/cilium/cilium/blob/master/pkg/envoy/server.go#L557-L593.
Feel free to factorize that code somewhere else.